### PR TITLE
Minor corrections for Make in 2021.

### PR DESCRIPTION
**Pull request by octonion**
Some minor issues building that were easy to correct.

### CODE REVIEW

File: paper/break_all_test.cc [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include "ibuckets.h"
 #include "test.h"
 #include "trie.h"
+#include "cstring"
```
Comment (danvk) [inline paper/break_all_test.cc]: Shouldn't this be `<cstring>`?

File: glog-src/config.h [hunk 2]
```diff
@@ -139,7 +139,7 @@
 #define PACKAGE_VERSION "0.3.2"
 
 /* How to access the PC from a struct ucontext */
-#define PC_FROM_UCONTEXT uc_mcontext->__ss.__rip
+/*#define PC_FROM_UCONTEXT uc_mcontext->__ss.__rip*/
```
Comment (danvk) [inline glog-src/config.h:142]: What's going on here? Is this an upstream change to glog?

File: Makefile [hunk 3]
```diff
@@ -1,5 +1,6 @@
 CC = g++
-CPPFLAGS = -g -Wall -O3 -I. -Iglog-src -Wno-sign-compare
+CPPFLAGS = -g -Wall -O3 -I. -Iglog-src -Wno-sign-compare -fpermissive
```
Comment (danvk) [inline Makefile]: What's the error you get without `-fpermissive`?

File: neighbors.cc [hunk 4]
```diff
@@ -16,6 +16,7 @@
 #include <iostream>
 #include "gflags/gflags.h"
 #include "init.h"
+#include <cstring>
```
Comment (danvk) [inline neighbors.cc]: Please keep library `#include`s grouped at the top of the list.

**Review by danvk — COMMENTED** [review]
Thanks for the change. I'm mostly curious about some details (and how you found this and what you're doing with it!).

File: paper/break_all_test.cc (hunk 1 above)
Comment (octonion) [reply paper/break_all_test.cc]: I just bought a Xeon Phi workstation, and I recalled (incorrectly!) that your Boggle code was multithreaded. It must have been some code that I wrote that was multithreaded, but it only took a little patching for your code to build successfully on both the Xeon Phi workstation and a more standard Xeon workstation. My apologies, these weren't the best patches, only the simplest that fixed the few build issues.

File: glog-src/config.h (hunk 2 above)
Comment (octonion) [reply glog-src/config.h:142]: It seems to be a change to glog sometime over the years - I checked the latest glog source code.

File: neighbors.cc (hunk 4 above)
Comment (octonion) [reply neighbors.cc]: Thanks, I'll make that change.

File: Makefile (hunk 3 above)
Comment (octonion) [reply Makefile]: It looks like the errors occur in glog-src/demangle.cc - "error: ISO C++ forbids comparison between pointer and integer [-fpermissive]". I'm merging in the upstream changes.

File: glog-src/demangle.cc [hunk 5]
```diff
@@ -167,7 +167,7 @@ static size_t StrLen(const char *str) {
 // Returns true if "str" has at least "n" characters remaining.
 static bool AtLeastNumCharsRemaining(const char *str, int n) {
   for (int i = 0; i < n; ++i) {
-    if (str == '\0') {
+    if (str[i] == '\0') {
```
Comment (danvk) [inline glog-src/demangle.cc:170]: Looks like g++ was right :)

File: Makefile [hunk 6]
```diff
@@ -1,5 +1,6 @@
 CC = g++
 CPPFLAGS = -g -Wall -O3 -I. -Iglog-src -Wno-sign-compare
```
Comment (danvk) [inline Makefile:2]: Does CPPFLAGS also need `-pthread`? https://stackoverflow.com/questions/15367617/creating-makefile-with-pthreads

**Review by danvk — APPROVED** [review]
Thanks for updating glog! I have to admit, I'm not sure I follow what `-pthread` is doing for the build here. The boggle-solving code is actually pretty hostile to fine-grained parallelization since the logic for checking if you've already found a word involves mutating the Trie https://github.com/danvk/performance-boggle/blob/156e7ade7c98490cc7435321f41886030ed7e836/4x4/boggler.cc#L28-L29

If you want parallelism, you need to have multiple Tries and multiple Bogglers.

File: glog-src/demangle.cc (hunk 5 above)
Comment (octonion) [reply glog-src/demangle.cc:170]: glog throws many "undefined reference to `pthread_rwlock_init'" without -pthread, so it doesn't seem to be an issue with your code at all.

**Comment by danvk** [comment]
Thanks @octonion!

**Pull request merged**